File: src/com/sterndu/multicore/OneTaskMultipleData.java
package com.sterndu.multicore;

import java.util.*;
import java.util.Map.Entry;
import com.sterndu.multicore.MultiCore.TaskHandler;
import com.sterndu.util.interfaces.ThrowingConsumer;

public class OneTaskMultipleData<T, E, O> extends TaskHandler {

	@FunctionalInterface
	public interface Task<E, O> {

		O run(E[] e);

	}
	private final Task<E, O> task;
	private final Map<T, O> results;
	private final List<Entry<T, E[]>> params_list;

	private final Object lock = new Object();
	private int active_tasks = 0;

	public OneTaskMultipleData(Task<E, O> task) {
		this.task = task;
		results = new HashMap<>();
		params_list = new LinkedList<>();
	}

	public OneTaskMultipleData(Task<E, O> task,double prioMult) {
		super(prioMult);
		this.task = task;
		results = new HashMap<>();
		params_list = new LinkedList<>();
	}

	private Entry<T,E[]> getParamsFromList(){
		if (params_list.size()>0) synchronized (params_list) {
			synchronized (lock) {
				active_tasks++;
			}
			Entry<T,E[]> e=params_list.get(0);
			params_list.remove(0);
			return e;
		}
		else return null;
	}

	private void putResult(T key, O res) {
		synchronized (lock) {
			active_tasks--;
		}
		results.put(key, res);
	}

	@Override
	protected boolean hasTask() {
		return params_list.size()>0;
	}

	public Map<T, O> getResults() {

		while (params_list.size() > 0 || active_tasks != 0) try {
			Thread.sleep(2);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
		return results;
	}

	@Override
	public ThrowingConsumer<TaskHandler> getTask() { return t->{
		Entry<T,E[]> entry=getParamsFromList();
		Task<E,O> ta=task;
		O res=ta.run(entry.getValue());
		putResult(entry.getKey(), res);
	}; }

	@SuppressWarnings("unchecked")
	public void pushParams(T key, E... e) {
		synchronized (this.params_list) {
			this.params_list.add(Map.entry(key, e));
		}
	}

}

File: src/com/sterndu/multicore/OneTaskMultipleData.kt
@file:JvmName("OneTaskMultipleData")
package com.sterndu.multicore

import java.util.*

class OneTaskMultipleData<T, E, O> : TaskHandler {
	fun interface Task<E, O> {
		fun run(e: Array<out E>): O
	}

	private val taskToBeRun: Task<E, O>
	private val results: MutableMap<T, O>
	private val paramsList: MutableList<Pair<T, Array<out E>>>
	private val lock = Any()
	private var activeTasks = 0

	constructor(
		task: Task<E, O>,
		millis: Long = 0,
		isFixedDelay: Boolean = false,
		nextRun: Long = 0,
		canRunSimultaneously: Boolean = false,
	): super(millis, isFixedDelay, nextRun, canRunSimultaneously) {
		this.taskToBeRun = task
		results = HashMap()
		paramsList = LinkedList()
	}

	private val paramsFromList: Pair<T, Array<out E>>?
		get() {
			if (paramsList.isNotEmpty()) synchronized(paramsList) {
				synchronized(lock) { activeTasks++ }
				return paramsList.removeAt(0)
			} else return null
		}

	private fun putResult(key: T, res: O) {
		synchronized(lock) { activeTasks-- }
		results[key] = res
	}

	override fun hasTask(): Boolean {
		return paramsList.isNotEmpty()
	}

	fun getResults(): Map<T, O> {
		while (paramsList.isNotEmpty() || activeTasks != 0) try {
			Thread.sleep(2)
		} catch (e: InterruptedException) {
			e.printStackTrace()
		}
		return results
	}

	override fun getTask(): (() -> Unit)? {
		val entry = paramsFromList
		return if (entry != null) {
			{
				val ta = taskToBeRun
				val res = ta.run(entry.second)
				putResult(entry.first, res)
			}
		} else null
	}

	fun pushParams(key: T, vararg e: E) {
		synchronized(paramsList) { paramsList.add(key to e) }
	}
}